upgrade ray-tracing branch to main branch

Vecvec opened the pull request:
**Connections**
upgrades this branch to wgpu trunk, also fixes the tlas memory leak for me

**Description**
This branch is missing a lot of commits, this PR adds those, and also brings the tlas and blas structs up to work with those changes.

**Testing**
Tested with the three examples added by this change, as well as a basic program that (in a loop) creates blas and tlas, and immediately frees them.

<!-- 
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown` (done by github actions)
  - [x] `--target wasm32-unknown-emscripten` (done by github actions)
- [x] Run `cargo xtask test` to run tests. (done by github actions)
- Add change to `CHANGELOG.md`. I do not think this is applicable as this is part of the ray-tracing PR (gfx-rs/wgpu#3631).

Now ready for review

jimblandy committed:
[naga] Introduce `ScalarKind::AbstractInt` and `AbstractFloat`.

Introduce new variants of `naga::ScalarKind`, `AbstractInt` and
`AbstractFloat`, for representing WGSL abstract types.

jimblandy committed:
[naga] Introduce `Literal::AbstractInt` and `AbstractFloat`.

Introduce new variants of `naga::Literal`, `AbstractInt` and
`AbstractFloat`, for representing WGSL abstract values.

jimblandy committed:
[naga wgsl-in] Reformat match statement for better patch stability.

The large `match` statement in `Lowerer::construct` seems to flop back
and forth between two indentation levels as it's edited, making the
diffs hard to read. Rewrite it to use deferred initialization of
`expr`, so that `cargo fmt` doesn't have to decide whether or not to
put the `match` on the same line as `let expr`. This makes subsequent
diffs easier to read.

jimblandy committed:
[wgsl-in]: Remove `Components::Many::first_component_ty_inner`.

Delete the `first_component_ty_inner` field from
`front::wgsl::lower::construction::Components::Many`. With the
introduction of abstract types, it will no longer be possible to infer
the type of the vector being constructed by looking at the type of its
first constructor argument alone: automatic conversion rules might
need to be applied to that argument.

jimblandy committed:
[naga wgsl-in] Implement abstract types for consts, constructors.

jimblandy committed:
[naga wgsl-in] Automatic conversions for global `var` initializers.

jimblandy committed:
[naga] Rename abstract-types snapshots abstract-types-const.

This prepares for introducing a similar test for global variables.

jimblandy committed:
[naga wgsl-in] Automatic conversions for local `var` initializers.

jimblandy committed:
[naga] Improve ConstantEvaluatorError::InvalidCastArg message.

torokati44 committed:
Bump web-sys to 0.3.65 (#4777)

cwfitzgerald committed:
Fixes for Skybox and All Examples (#4780)

gents83 committed:
Remove resources ONLY when needed inside wgpu and not in user land (#4782)

gents83 committed:
Don't keep a strong ref in storage for destroyed resources (#4786)

jimblandy committed:
[naga wgsl-in] Clarify match in `automatic_conversion_join`.

Co-authored-by: Teodor Tanasoaia <28601907+teoxoy@users.noreply.github.com>

jimblandy committed:
[naga] Improve snapshot output when validation fails.

jimblandy committed:
[naga wgsl-in] Matrix constructors have only float overloads.

dependabot[bot] committed:
Bump js-sys from 0.3.65 to 0.3.66

Bumps [js-sys](https://github.com/rustwasm/wasm-bindgen) from 0.3.65 to 0.3.66.
- [Release notes](https://github.com/rustwasm/wasm-bindgen/releases)
- [Changelog](https://github.com/rustwasm/wasm-bindgen/blob/main/CHANGELOG.md)
- [Commits](https://github.com/rustwasm/wasm-bindgen/commits)

---
updated-dependencies:
- dependency-name: js-sys
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

exrook committed:
Delete root.rs so git recognizes rename in following commit

exrook committed:
Rename gpu.rs to root.rs

a1phyr committed:
Fix Arcanization changelog (#4795)

xiaopengli89 committed:
Support nv12 texture format (#4573)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

dependabot[bot] committed:
Bump winit from 0.29.3 to 0.29.4 (#4784)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jimblandy committed:
[naga wgsl-in] Constructors with types don't make abstract values.

When a constructor builtin has an explicit type parameter, like
`mat2x2<f32>`, it should not produce an abstract matrix, even if its
arguments are abstract.

dependabot[bot] committed:
Bump JamesIves/github-pages-deploy-action from 4.4.3 to 4.5.0 (#4798)

Bumps [JamesIves/github-pages-deploy-action](https://github.com/jamesives/github-pages-deploy-action) from 4.4.3 to 4.5.0.
- [Release notes](https://github.com/jamesives/github-pages-deploy-action/releases)
- [Commits](https://github.com/jamesives/github-pages-deploy-action/compare/v4.4.3...v4.5.0)

---
updated-dependencies:
- dependency-name: JamesIves/github-pages-deploy-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump web-sys from 0.3.65 to 0.3.66 (#4800)

Bumps [web-sys](https://github.com/rustwasm/wasm-bindgen) from 0.3.65 to 0.3.66.
- [Release notes](https://github.com/rustwasm/wasm-bindgen/releases)
- [Changelog](https://github.com/rustwasm/wasm-bindgen/blob/main/CHANGELOG.md)
- [Commits](https://github.com/rustwasm/wasm-bindgen/commits)

---
updated-dependencies:
- dependency-name: web-sys
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump wasm-bindgen-futures from 0.4.38 to 0.4.39 (#4799)

Bumps [wasm-bindgen-futures](https://github.com/rustwasm/wasm-bindgen) from 0.4.38 to 0.4.39.
- [Release notes](https://github.com/rustwasm/wasm-bindgen/releases)
- [Changelog](https://github.com/rustwasm/wasm-bindgen/blob/main/CHANGELOG.md)
- [Commits](https://github.com/rustwasm/wasm-bindgen/commits)

---
updated-dependencies:
- dependency-name: wasm-bindgen-futures
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

teoxoy committed:
rename `DeviceDescriptor.features` to `DeviceDescriptor.required_features`

teoxoy committed:
rename `DeviceDescriptor.limits` to `DeviceDescriptor.required_limits`

teoxoy committed:
improve features/limits docs

teoxoy committed:
move `alignments` & `downlevel` `.clone()`s inside function

teoxoy committed:
add changelog entry

Wumpf committed:
Remove surface extent validation (and thus fix the annoying `Requested size ... is outside of the supported range` warning) (#4796)

* Remove surface extent validation

* silence pnext vulkan validation warning which can happen on surface resize

* remove old VUID-VkSwapchainCreateInfoKHR-imageExtent-01689 validation warning ignore

* Validate surface against max texture size

ComfyFluffy committed:
Add space to a comment inside make_spirv_raw (#4804)

jimblandy committed:
[naga wgsl-in] Drop spanless labels from front-end error messages.

When a label in a WGSL front end error has an undefined span, omit the
label from the error message. This is not great, but because of the
way Naga IR represents local variable references it is hard to get the
right span, and omitting the label better than panicking in `unwrap`,
since the error message has a general message anyway.

jimblandy committed:
[naga wgsl-in] Use a better span for errors in constructors.

When reporting errors in construction expressions, use the span of the
constructor itself (that is, the type name) in preference to the span
of the overall expression. This makes errors easier to follow.

crowlKats committed:
update deno (#4801)

daniel-keitel committed:
implemented requested changes

daniel-keitel committed:
Merge remote-tracking branch 'origin/master' into hal-acceleration-structures

Blatko1 committed:
Fixes and changes to the documentation for increasing clarity (#4806)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

Vecvec committed:
Merge branch 'ray-tracing' into trunk

# Conflicts:
#	Cargo.lock
#	tests/Cargo.toml
#	wgpu-core/src/binding_model.rs
#	wgpu-core/src/command/compute.rs
#	wgpu-core/src/command/mod.rs
#	wgpu-core/src/command/render.rs
#	wgpu-core/src/device/life.rs
#	wgpu-core/src/device/queue.rs
#	wgpu-core/src/device/resource.rs
#	wgpu-core/src/hub.rs
#	wgpu-core/src/resource.rs
#	wgpu-core/src/track/mod.rs
#	wgpu-hal/Cargo.toml
#	wgpu-hal/src/dx11/mod.rs
#	wgpu-hal/src/lib.rs
#	wgpu-hal/src/vulkan/adapter.rs
#	wgpu-types/src/lib.rs
#	wgpu/Cargo.toml
#	wgpu/src/context.rs

Vecvec committed:
fix unexpected build errors (not the expected ones)

Vecvec committed:
few more patches

Vecvec committed:
Merge remote-tracking branch 'pull_Vecvec/hal-acceleration-structures' into ray-tracing

# Conflicts:
#	Cargo.lock
#	wgpu-hal/Cargo.toml
#	wgpu-hal/src/dx11/mod.rs
#	wgpu-types/src/lib.rs

teoxoy committed:
fix expected error message

almarklein committed:
Add feature float32-filterable (#4759)

dependabot[bot] committed:
Bump core-graphics-types from 0.1.2 to 0.1.3 (#4813)

dependabot[bot] committed:
Bump wasm-bindgen-test from 0.3.38 to 0.3.39 (#4807)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

nical committed:
Use Display instead of Debug to log errors. (#4830)

xiaopengli89 committed:
check multi-planar format

teoxoy committed:
[vk] remove (old) unused blocklist for dual source blending (#4758)

valaphee committed:
Remove DX11 backend (#4828)

valaphee committed:
[gl] add support for line and point polygon modes (#4836)

Co-authored-by: Nicolas Silva <nical@fastmail.com>

dependabot[bot] committed:
Bump futures-lite from 2.0.1 to 2.1.0 (#4820)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

daxpedda committed:
Expose shader validation (#4811)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

daniel-keitel committed:
[wgpu-hal] Inline RayQuery Support (#3507)

Co-authored-by: JMS55 <47158642+JMS55@users.noreply.github.com>
Co-authored-by: Ashley Ruglys <ashley.ruglys@gmail.com>
Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

DouglasDwyer committed:
Fix Javascript exception on repeated `BufferSlice::get_mapped_range` calls (#4726)

* Add reusable buffer mappings for WASM

* Run cargo fmt

* Update CHANGELOG.md

* Update web.rs

* Add documentation for WebBuffer struct

cwfitzgerald committed:
Remove `expose-ids` Feature (#4841)

* Remove expose-ids feature

* Changelog

cwfitzgerald committed:
Some minor cleanups (#4843)

nical committed:
Work around cbindgen issue

cbindgen spins out of control if wgpu_hal reexports a wgpu_types type using pub type in stead of pub use while keeping the same name.

teoxoy committed:
[spv-out] Decorate all non-uniform binding array access chains as `NonUniform`, not just buffer binding arrays.

Apply the `NonUniform` decoration to the results of all access chains rooted in binding arrays that use non-uniform values as indices, regardless of the binding array's element type and address space. Previously, Naga only decorated non-uniform access chains for binding arrays of buffers.

teoxoy committed:
Update multi-planar texture API (#4837)

TheoDulka committed:
Use `VK_EXT_robustness2` only when not using an outdated intel iGPU driver (#4602)

Vecvec committed:
fix more compile errors

Vecvec committed:
fix err

Vecvec committed:
Merge branch 'trunk' into ray-tracing

# Conflicts:
#	wgpu-core/src/binding_model.rs
#	wgpu-core/src/device/resource.rs
#	wgpu-hal/examples/ray-traced-triangle/main.rs
#	wgpu-hal/src/dx11/command.rs
#	wgpu-hal/src/dx11/device.rs
#	wgpu-hal/src/dx11/mod.rs
#	wgpu-hal/src/lib.rs
#	wgpu-types/src/lib.rs

dependabot[bot] committed:
Bump once_cell from 1.18.0 to 1.19.0 (#4847)

Bumps [once_cell](https://github.com/matklad/once_cell) from 1.18.0 to 1.19.0.
- [Changelog](https://github.com/matklad/once_cell/blob/master/CHANGELOG.md)
- [Commits](https://github.com/matklad/once_cell/compare/v1.18.0...v1.19.0)

---
updated-dependencies:
- dependency-name: once_cell
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jimblandy committed:
[naga wgsl-in] Relocate `automatic_conversion_consensus`.

Move `naga::front::wgsl::lower::construction::automatic_conversion_consensus`
into `conversion.rs`, and make it a method of `ExpressionContext`.

jimblandy committed:
[naga wgsl-in] Relax arg type for `automatic_conversion_consensus`.

jimblandy committed:
[naga wgsl-in] Rename `convert_slice_to_common_leaf_scalar`.

This is a mouthful, but it's going to align better with functions to
be introduced in later commits.

jimblandy committed:
[naga wgsl-in] Separate out `convert_to_leaf_scalar`.

Abstract out the body of `convert_slice_to_common_leaf_scalar`'s main
loop into its own function. Code motion only, no intended change of
behavior.

Wumpf committed:
Fix `BufferMapRange` not being Send/Sync (on native) (#4818)

* Make `BufferMappedRange` trait WasmNotSendSync

* make QueueWriteBuffer trait Debug

* add changelog

* fix pr numbers in changelog

* missing send/sync on web.rs on BufferMappedRange for fragile-send-sync-non-atomic-wasm + !atomics

nical committed:
Format extent and origin types as tuples. (#4853)

* Format extent and origin types as tuples.

* Update wgpu-types/src/lib.rs

Co-authored-by: Andreas Reich <r_andreas2@web.de>

---------

Co-authored-by: Andreas Reich <r_andreas2@web.de>

valaphee committed:
[wgpu] correct doc for resolve query set (#4849)

* Correct doc for resolve query set

* Add note about wgpu::QUERY_SIZE

---------

Co-authored-by: Andreas Reich <r_andreas2@web.de>

Vecvec committed:
fix all compile errors

Vecvec committed:
Merge pull request #4 from gfx-rs/trunk

upgrade

Vecvec committed:
fix clippy errors

Vecvec committed:
Merge remote-tracking branch 'origin/ray-tracing' into ray-tracing

Vecvec committed:
fix issues in ray-traced-triangle example

Vecvec committed:
fix clippy warnings

Vecvec committed:
fix wasm not compiling

Vecvec committed:
fix test

Toqozz committed:
Add `COPY_SRC` to Metal's surface usage bits. (#4852)

Vecvec committed:
fix clippy (again)

Vecvec committed:
fix some rustfmt issues

Vecvec committed:
fix more rustfmt issues

Vecvec committed:
fix even more rustfmt issues

jimblandy committed:
[naga wgsl-in] Support abstract operands to binary operators.

Vecvec committed:
fix yet more rustfmt issues

dependabot[bot] committed:
Bump libc from 0.2.150 to 0.2.151 (#4854)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump profiling from 1.0.11 to 1.0.12 (#4855)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump tokio from 1.34.0 to 1.35.0 (#4856)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Vecvec committed:
add #[allow(dead_code)] to some structs to fix clippy on web assembly

Vecvec committed:
more format fixes

Vecvec committed:
even more format fixes

Vecvec committed:
close to final format fixes

Vecvec committed:
final format fixes (hopefully)

Vecvec committed:
actual final format fixes (hopefully)

Vecvec committed:
Merge pull request #5 from gfx-rs/trunk

merge with gfx wgpu again

teoxoy committed:
[glsl-in] fix swizzle in a global const context

jimblandy committed:
[naga wgsl-in] Improve error message for failed concretization.

jimblandy committed:
[naga] Change operators.wgsl to use values that matter.

Change `splat` to take run-time values as arguments, to prevent
constant evaluation from boiling everything away.

jimblandy committed:
[naga] Teach the constant evaluator vector/vector operators.

Allow constant evaluation of binary operators whose left and right
operands are both vectors.

Vecvec committed:
move ray_cube_cube example and fix bugs

Vecvec committed:
move other examples

Vecvec committed:
fix crash when dropping

Vecvec committed:
fix test

Vecvec committed:
apply clippy and rustfmt changes

dependabot[bot] committed:
Bump syn from 2.0.39 to 2.0.40 (#4864)

Bumps [syn](https://github.com/dtolnay/syn) from 2.0.39 to 2.0.40.
- [Release notes](https://github.com/dtolnay/syn/releases)
- [Commits](https://github.com/dtolnay/syn/compare/2.0.39...2.0.40)

---
updated-dependencies:
- dependency-name: syn
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jimblandy committed:
[naga wgsl-out] Include the `i` suffix on `i32` literals. (#4863)

Without the suffix, `Expression::Literal(Literal::I32)` expressions
get written without any suffix on the decimal number, meaning that
they get re-parsed as AbstractInt values. In theory, this should
always be fine, but since we don't actually support abstract types yet
in all the places we should, having them appear in the output causes
validation problems.

xiaopengli89 committed:
fix present mode for wgl (#4865)

jimblandy committed:
[naga wgsl-out] Include the `f` suffix on `f32` literals.

Without the suffix, `Expression::Literal(Literal::F32)` expressions
get written without any suffix on the number, meaning that they get
re-parsed as `AbstractFloat` values. In theory, this should always be
fine, but since we don't actually support abstract types yet in all
the places we should, having them appear in the output causes
validation problems.

See also: #4863, which did the same for `i32` literals.

jimblandy committed:
[naga wgsl] Let unary operators accept and produce abstract types.

Fixes #4445.
Fixes #4492.
Fixes #4435.

jimblandy committed:
[naga-cli] Abstract out parsing of input file.

jimblandy committed:
[naga-cli] Add `--bulk-validate` option.

jimblandy committed:
[naga xtask] Use naga-cli's --bulk-validate option.

Use `naga --bulk-validate` for `cargo xtask validate wgsl`, reducing
runtime from 12s to 0.8s.

Vecvec committed:
fix some memory leaks

dependabot[bot] committed:
Bump ctor from 0.2.5 to 0.2.6 (#4873)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump syn from 2.0.40 to 2.0.41 (#4874)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Vecvec committed:
Merge from gfx-rs/trunk

get up to latest

nical committed:
Reintroduce buffer snatching Part 1 (#4867)

Introduce snatch guards

dependabot[bot] committed:
Bump zerocopy from 0.7.26 to 0.7.31 (#4882)

Bumps [zerocopy](https://github.com/google/zerocopy) from 0.7.26 to 0.7.31.
- [Release notes](https://github.com/google/zerocopy/releases)
- [Changelog](https://github.com/google/zerocopy/blob/main/CHANGELOG.md)
- [Commits](https://github.com/google/zerocopy/compare/v0.7.26...v0.7.31)

---
updated-dependencies:
- dependency-name: zerocopy
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

nical committed:
Refactor create_buffer so that we can snatch the raw buffer in the error path. (#4878)

The general idea is to register postpone reigistering the buffer until towards the end of the function so that our unique reference to it lets us easily snatch the raw buffer if an error happens.

daxpedda committed:
Add `wgpu` crate features for backends (#4815)

* Introduce `dx12` and `metal` crate features to `wgpu`

* Implement dummy `Context` to allow compilation with `--no-default-features`

* Address review

* Remove `dummy::Context` in favor of `hal::api::Empty`

* Add changelog entry

* Panic early in `Instance::new()` if no backend is enabled

Co-Authored-By: Andreas Reich <1220815+Wumpf@users.noreply.github.com>

---------

Co-authored-by: Andreas Reich <1220815+Wumpf@users.noreply.github.com>

nical committed:
Allow clippy::pattern_type_mismatch (#4887)

Vecvec committed:
Merge remote-tracking branch 'gfx-rs/trunk' into ray-tracing

# Conflicts:
#	wgpu-core/src/device/resource.rs

Vecvec committed:
fix incompatibility with new wgpu trunk changes

Vecvec committed:
fix clippy & fmt

Vecvec committed:
fix clippy errors on github

Wumpf committed:
Document wgpu & wgpu-core features (#4886)

scoopr committed:
validation: More detailed on incompatible BGL (#4826)

daxpedda committed:
WebGL: add support for more `RawWindowHandle` variants (#4888)

dependabot[bot] committed:
Bump thiserror from 1.0.50 to 1.0.51 (#4890)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

nical committed:
Simplify `ResourceMaps` (#4880)

* Remove the abstractions in resource maps

ResourceMaps had a rather convoluted system for erasing types that isn't needed anywhere except in one place in the triage_resource function. Everywhere else we are always dealing with specific types so using a member of the resource maps is simpler than going through an abstraction. More importantly there was a constraint that all contents of the resource maps implement the Resource trait which got in the way of some of the ongoing buffer snatching changes.

This commit simplifies this by removing the abstraction. Each resource type has its hash map directly in ResourceMaps and it is easier to have different requirements and behaviors depending on the type of the each resource.

nical committed:
Make the naga version in trunk as high as the latest published one. (#4891)

nical committed:
Avoid allocating memory every time we might log a label (#4893)

We allocate a String every time we want to get a label for logging. The string is also allocated when logging is disabled. Either way, the allocation is unnecessary. This commit replaces the String with a dyn Debug reference which does not need any allocation.

Vecvec committed:
Merge remote-tracking branch 'gfx-rs/trunk' into ray-tracing

# Conflicts:
#	wgpu-core/src/device/life.rs

Vecvec committed:
update to trunk

Vecvec committed:
remove a line that printed log "Created blas ..." in device create tlas

jimblandy committed:
[naga xtask] Move validation subcommands into their own module.

nical committed:
Remove some locks in BindGroup (#4894)

* Remove some locks in BindGroup

These are only written to clear the vectors when triaging bindgroups for destruction, which is not necessary. We can let the reference counts drop when the bind group is dropped.

* Make the mem_leak test pass again

jimblandy committed:
[naga xtask] Break validation subcommands into functions.

jimblandy committed:
[naga xtask] Produce output only when an error occurs.

jimblandy committed:
[naga xtask] Collect validation jobs in a Vec before running them.

jimblandy committed:
[naga xtask] Remove `ack_visiting` function.

jimblandy committed:
[naga xtask] Use `indicatif` crate to report validation progress.

jimblandy committed:
[naga xtask] Run validation jobs in parallel, using jobserver.

jimblandy committed:
[naga xtask] Add `validate all` subcommand.

Commit:
[wgpu-core] Ensure that DeviceLostCallback is always called exactly once

* Ensure device lost closure is called exactly once before being dropped.

This requires a change to the Rust callback signature, which is now Fn
instead of FnOnce. When the Rust callback or the C closure are dropped,
they will panic if they haven't been called. `device_drop` is changed
to call the closure with a message of "Device dropped." A test is added.

sotaroikeda committed:
Support Device fence sharing with dx12 on Windows (#4900)

Co-authored-by: Sotaro Ikeda <you@example.com>

dependabot[bot] committed:
Bump tokio from 1.35.0 to 1.35.1 (#4903)

Bumps [tokio](https://github.com/tokio-rs/tokio) from 1.35.0 to 1.35.1.
- [Release notes](https://github.com/tokio-rs/tokio/releases)
- [Commits](https://github.com/tokio-rs/tokio/compare/tokio-1.35.0...tokio-1.35.1)

---
updated-dependencies:
- dependency-name: tokio
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

nical committed:
Check that raw buffers and raw bind groups are valid (#4895)

miaobuao committed:
fix: docs (#4901)

nical committed:
Simplify some code around buffer unmapping (#4892)

Vecvec committed:
Merge remote-tracking branch 'gfx-rs/trunk' into ray-tracing

# Conflicts:
#	wgpu-core/src/device/life.rs

nical committed:
Buffer snatching (#4896)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

Commit:
Eagerly release GPU resources when we lose the device. (#4851)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

cwfitzgerald committed:
Use nightly for docs (#4906)

dependabot[bot] committed:
Bump syn from 2.0.41 to 2.0.42 (#4908)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump profiling from 1.0.12 to 1.0.13 (#4909)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump anyhow from 1.0.75 to 1.0.76 (#4910)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

cwfitzgerald committed:
Remove id32 feature

cwfitzgerald committed:
Add BGL Deduplication Index Test (#4914)

Vecvec committed:
Merge remote-tracking branch 'gfx-rs/trunk' into ray-tracing

# Conflicts:
#	wgpu-core/src/device/life.rs

HactarCE committed:
Fix typo "layout pipeline layout" -> "pipeline layout" (#4921)

ErichDonGubler committed:
docs: inline `document-features` usage, remove dep.

Re-implements https://github.com/gfx-rs/wgpu/pull/4886 (CC @Wumpf)
without the `document-features` crate, which has issues integrating into
Firefox builds after being `cargo vendor`ed into its repository. This
issue is being tracked against
https://github.com/slint-ui/document-features/issues/20. Once resolved,
I expect that we will want to revert this PR in its entirety, since
`document-features` is still a good addition to `wgpu`'s documentation
story.

Internally, consensus has already been achieved for this change.
Firefox's ability to build unfortunately take priority over this
particular convenience. Hopefully, we won't have to compromise shortly!

I tested this by ensuring that the HTML output of our existing
`document_features::document_features!(…)` usage was exactly the same.
There should be exactly zero regressions in the current state of
documentation for users. For maintainers, I have added a disclaimer that
one needs to keep changes in sync. with the relevant `Cargo.toml`
manifests.

dependabot[bot] committed:
Bump winit from 0.29.4 to 0.29.5 (#4922)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump winit from 0.29.5 to 0.29.6 (#4930)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump web-time from 0.2.3 to 0.2.4 (#4929)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jimblandy committed:
[naga xtask] Use `log::error` in preference to `eprintln`.

jimblandy committed:
[naga xtask] Use `anyhow::ensure!` where appropriate.

ErichDonGubler committed:
[naga xtask] Add and use `ValidateSubcommand::all`.

Add an `all` method to `ValidateSubcommand`, so that we can just use
matches elsewhere, and let the compile catch missing or duplicated
cases for us.

littledivy committed:
Align `wgpu_types::CompositeAlphaMode` serde serializations to spec (#4940)

* Align wgpu_types::CompositeAlphaMode serde serializations to spec

* add changelog

dependabot[bot] committed:
Bump anyhow from 1.0.76 to 1.0.77 (#4936)

Bumps [anyhow](https://github.com/dtolnay/anyhow) from 1.0.76 to 1.0.77.
- [Release notes](https://github.com/dtolnay/anyhow/releases)
- [Commits](https://github.com/dtolnay/anyhow/compare/1.0.76...1.0.77)

---
updated-dependencies:
- dependency-name: anyhow
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump thiserror from 1.0.51 to 1.0.52 (#4933)

Bumps [thiserror](https://github.com/dtolnay/thiserror) from 1.0.51 to 1.0.52.
- [Release notes](https://github.com/dtolnay/thiserror/releases)
- [Commits](https://github.com/dtolnay/thiserror/compare/1.0.51...1.0.52)

---
updated-dependencies:
- dependency-name: thiserror
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump syn from 2.0.42 to 2.0.43 (#4932)

Bumps [syn](https://github.com/dtolnay/syn) from 2.0.42 to 2.0.43.
- [Release notes](https://github.com/dtolnay/syn/releases)
- [Commits](https://github.com/dtolnay/syn/compare/2.0.42...2.0.43)

---
updated-dependencies:
- dependency-name: syn
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

JMS55 commented:
@Vecvec I would love to have this in wgpu. Given that @daniel-keitel seems to be busy atm, perhaps you want to open your own PR in the wgpu repo? (ofc crediting daniel-keitel).

Vecvec committed:
Merge remote-tracking branch 'gfx-rs/trunk' into ray-tracing

nical committed:
Snatch the raw texture when destroying it

nical committed:
Put raw texture access behind snatch guards

cwfitzgerald committed:
BGL Weak Pointer Deduplication Pool (#4927)

* Ho boy

* BGL pool finished

* Remove id32 feature

* Add BGL Test

* Iteration

* Working Dedupe

* Tests

* Iteration

* Re-iteration

* Hash Cleanup

* Add large slew of tests

* Whoops

Dinnerbone committed:
Change examples page menu on smaller screens (#4958)

dependabot[bot] committed:
Bump anyhow from 1.0.77 to 1.0.78 (#4950)

Bumps [anyhow](https://github.com/dtolnay/anyhow) from 1.0.77 to 1.0.78.
- [Release notes](https://github.com/dtolnay/anyhow/releases)
- [Commits](https://github.com/dtolnay/anyhow/compare/1.0.77...1.0.78)

---
updated-dependencies:
- dependency-name: anyhow
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

cwfitzgerald committed:
Disable Linux Tests (#4957)

Dinnerbone committed:
Fix incorrect ConfigureSurfaceError::TooLarge message (#4960)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

Vecvec commented:
Thank you! I had assumed @daniel-keitel was just away for Christmas and was not rushing. Currently I am away too and so cannot check merges with trunk have not broken anything, which I would like to do before opening a new PR. If @daniel-keitel is still away when I get back I will create a PR in wgpu as you suggest.

daxpedda committed:
Add `cfg_aliases` to `wgpu` (#4935)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

vilcans committed:
hello_compute: check for missing command-line args (#4939)

* hello_compute: check for missing command-line args

Fixes off-by-one error when checking for missing arguments.
Before this, running the example gave this scary looking error message:

```
$ cargo run --bin wgpu-examples hello_compute
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
     Running `target/debug/wgpu-examples hello_compute`
[2023-12-27T22:14:26Z ERROR wgpu::backend::direct] Handling wgpu errors as fatal by default
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In Device::create_bind_group
    Buffer binding size 0 is less than minimum 4
      note: buffer = `Storage Buffer`

', wgpu/src/backend/direct.rs:3139:5
```

As this was the first example I tried to run, it almost scared me
away, thinking it was a driver issue.

Instead, without arguments the example should use defaults.

* Add PR #4939 to changelog as instructed

dependabot[bot] committed:
Bump winit from 0.29.6 to 0.29.8 (#4948)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

rukai committed:
Fix xtask wasm-bindgen install (#4944)

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

teoxoy committed:
[gl] fix RGBA8 format capabilities (#4968)

- `Rgba8UnormSrgb` can't be used as a storage texture
- `Rgba8Snorm` can be used as a storage texture

nical committed:
Check thata bindgroup's textures are not destroyed before using it

nical committed:
Rename Texture::as_raw into Texture::raw

For consistency with Buffer::raw.

udoprog committed:
Avoid allocating during queue submit (#4947)

dependabot[bot] committed:
Bump serde from 1.0.193 to 1.0.194 (#4965)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

cwfitzgerald committed:
Fix hang in multithreaded test (#4975)

dependabot[bot] committed:
Bump anyhow from 1.0.78 to 1.0.79 (#4966)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

dependabot[bot] committed:
Bump thiserror from 1.0.52 to 1.0.56 (#4978)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump syn from 2.0.46 to 2.0.47 (#4979)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

cwfitzgerald committed:
Use Custom Mesa for Building (#4977)

dependabot[bot] committed:
Bump serde_json from 1.0.108 to 1.0.110 (#4981)

Bumps [serde_json](https://github.com/serde-rs/json) from 1.0.108 to 1.0.110.
- [Release notes](https://github.com/serde-rs/json/releases)
- [Commits](https://github.com/serde-rs/json/compare/v1.0.108...v1.0.110)

---
updated-dependencies:
- dependency-name: serde_json
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

ids1024 committed:
wgpu-hal: Fix Mesa version check for version with suffix containing `.` (#4959)

On Pop!_OS we have versions like
`Mesa 23.3.0-1pop0~1702935939~22.04~67e417a`. This failed to parse here
since it tried to split at the `.` in the suffix.

Not sure if other distros use a suffix with a `.`, but splitting from
the left and comparing as a tuple instead of a float seems cleaner
overall.

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

cwfitzgerald committed:
Shorten Lock Lifetimes (#4976)

cwfitzgerald committed:
Pin DXC and Vulkan SDK version (#4980)

valaphee committed:
gles: use already existing debug__fn private capabilty instead of checking extensions (#4974)

cwfitzgerald committed:
Remove Mac CI (#4987)

dtzxporter committed:
Fix Queue::write_texture, Fix DX12 write_texture_subset_2d and re-enable the test. (#4990)

dependabot[bot] committed:
Bump syn from 2.0.47 to 2.0.48 (#4994)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump serde_json from 1.0.110 to 1.0.111 (#4993)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump gpu-allocator from 0.24.0 to 0.25.0 (#4992)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

cwfitzgerald committed:
Update Rollup

cwfitzgerald committed:
Add Verbosity Flags to wgpu-info (#4995)

Patryk27 committed:
[naga wgsl-in] Fix parsing `break if`s

Closes https://github.com/gfx-rs/wgpu/issues/4982.

cwfitzgerald committed:
Pin nightly docs version

nical committed:
Propagate errors when openning/closing a command encoder (#4999)

dependabot[bot] committed:
Bump serde from 1.0.194 to 1.0.195 (#5005)

Bumps [serde](https://github.com/serde-rs/serde) from 1.0.194 to 1.0.195.
- [Release notes](https://github.com/serde-rs/serde/releases)
- [Commits](https://github.com/serde-rs/serde/compare/v1.0.194...v1.0.195)

---
updated-dependencies:
- dependency-name: serde
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Andreas Reich <r_andreas2@web.de>

dependabot[bot] committed:
Bump winit from 0.29.8 to 0.29.9 (#5004)

Bumps [winit](https://github.com/rust-windowing/winit) from 0.29.8 to 0.29.9.
- [Release notes](https://github.com/rust-windowing/winit/releases)
- [Changelog](https://github.com/rust-windowing/winit/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rust-windowing/winit/compare/v0.29.8...v0.29.9)

---
updated-dependencies:
- dependency-name: winit
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Andreas Reich <r_andreas2@web.de>

rbran committed:
fix: get CONTEXT_PROFILE_MASK in invalid version (#5007)

Vecvec committed:
Merge remote-tracking branch 'gfx-rs/trunk' into ray-tracing

# Conflicts:
#	wgpu-core/src/device/life.rs

JMS55 commented:
No worries, thanks! Also, I'd love to see BLAS compaction support if you feel like implementing it. It's a major missing part of RT.

Vecvec committed:
fix merge issues

Vecvec commented:
I'd love to implement this, however looking at Vulkan specs [VkCopyAccelerationStructureInfoKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkCopyAccelerationStructureInfoKHR) it seems like this needs another hal function, because the suggested specification says :
> [Function compact_blas] Creates a new blas and copies (in a compacting way) the contents of the provided blas into the new one (compaction flag must be set). 

(https://github.com/gfx-rs/wgpu/issues/1040#issuecomment-1484274170)
Should something such as copy_acceleration_structure_to_acceleration_structure be implemented or is there some other way I have missed that does this?

JMS55 commented:
Yes, this will need some additional work in wgpu-hal. Should hopefully be pretty straightforward for that part, as wgpu-hal tends to be a simple wrapper over VK/DX/Metal. 
See https://developer.nvidia.com/blog/tips-acceleration-structure-compaction and https://nvpro-samples.github.io/vk_raytracing_tutorial_KHR/#accelerationstructure/bottom-levelaccelerationstructure

Vecvec commented:
Something like this for Vulkan?
````rust
unsafe fn copy_acceleration_structure_to_acceleration_structure<T>(&mut self, src: &Api::AccelerationStructure, dst: &Api::AccelerationStructure, copy: AccelerationStructureCopy) {
        let ray_tracing_functions = self
            .device
            .extension_fns
            .ray_tracing
            .as_ref()
            .expect("Feature `RAY_TRACING` not enabled");

        let structure_type = match copy.type_flags {
            AccelerationStructureType::Triangles => vk::StructureType::ACCELERATION_STRUCTURE_GEOMETRY_TRIANGLES_DATA_KHR,
            AccelerationStructureType::AABBs => vk::StructureType::ACCELERATION_STRUCTURE_GEOMETRY_AABBS_DATA_KHR,
            AccelerationStructureType::Instances => vk::StructureType::ACCELERATION_STRUCTURE_GEOMETRY_INSTANCES_DATA_KHR,
        };

        let mode = match copy.copy_flags {
            AccelerationStructureCopy::Clone => vk::CopyAccelerationStructureModeKHR::CLONE,
            AccelerationStructureCopy::Compact => vk::CopyAccelerationStructureModeKHR::COMPACT,
        };

        unsafe {
            ray_tracing_functions.acceleration_structure.copy_acceleration_structure(
                vk::DeferredOperationKHR::null(),
                &vk::CopyAccelerationStructureInfoKHR {
                    s_type: structure_type,
                    p_next: std::ptr::null(),
                    src: src.raw,
                    dst: dst.raw,
                    mode,
                }
            ).expect("Copy Failed");
        }
    }
````

JMS55 commented:
Something like that, yeah. I think you also need to add support for a BLAS size query, in order for the user to be able to get what size the compacted BLAS should be.

Vecvec commented:
And a size query like this?
````rust
unsafe fn get_acceleration_structure_compact_size(&self, acceleration_structure: &super::AccelerationStructure) -> wgt::BufferAddress {
        let ray_tracing_functions = self
            .shared
            .extension_fns
            .ray_tracing
            .as_ref()
            .expect("Feature `RAY_TRACING` not enabled");
        let mut size = vk::DeviceSize::default().to_ne_bytes();

        unsafe {
            ray_tracing_functions
                .acceleration_structure
                .write_acceleration_structures_properties(
                    &[acceleration_structure.raw],
                    vk::QueryType::ACCELERATION_STRUCTURE_COMPACTED_SIZE_KHR,
                    &mut size,
                    std::mem::size_of::<vk::DeviceSize>(),
                ).expect("getting properties failed");
        }
        wgt::BufferAddress::from_ne_bytes(size)
    }
````

dependabot[bot] committed:
Bump futures-lite from 2.1.0 to 2.2.0 (#5011)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump libc from 0.2.151 to 0.2.152 (#5010)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

JMS55 commented:
Probably. I've also just noticed there's a way to serialize and deserialize BLAS's, which is very cool. That would be nice to have too, although less necessary. You should probably talk to the wgpu devs for specific wpgu-related advice though, I'm not a maintainer, and haven't written much code for it.

cwfitzgerald committed:
Disable naga-cli's docs

daniel-keitel commented:
Hi, I will be fully back from my Christmas break tomorrow (with access to my desktop pc).
I will look over your changes this week.
After that we should, coordinate what we want to implement. (@Vecvec  I would preferer talking over Discord, if it is ok for you)
As far as I can tell a drastic change in the Api will be necessary to be idiomatic in reference to Webgpu.

nical committed:
Track the actual size of the buffer for memory init (#5014)

cwfitzgerald committed:
Re-Enable Mac CI (#5015)

Vecvec commented:
That's great! Hope you had a good holiday. 
I'm sorry but I don't have discord (and no intention of creating an account).
How drastic a change are you contemplating?

Vecvec commented:
I have also put the compact blas function onto a separate branch (vecvec:ray-tracing-compaction), as I don't think it's in the scope of this PR.

Vecvec committed:
Merge remote-tracking branch 'gfx-rs/trunk' into ray-tracing

jimblandy committed:
[wgpu-core] Minor doc fixes for initialization tracking. (#5017)

jimblandy committed:
Prefer `Arc::into_inner` over `Arc::try_unwrap`. (#5018)

nical committed:
Validate that a binding offset fits in the buffer

nical committed:
Validate max_vertex_buffers in render bundles. (#5025)

nical committed:
Don't case the dynamic offsets count to u8 (#5026)

nical committed:
Fix a bounds check in remove_abandoned (#5024)

dependabot[bot] committed:
Bump getrandom from 0.2.11 to 0.2.12 (#5030)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

nical committed:
Don't crash if a texture is destroyed before queue submission (#5028)

daniel-keitel commented:
@Vecvec the main change I am contemplating, is to somehow make the information stored in the creation info part of the wgpu-core library and reference count all required resources.

The problem at the moment is, that a reference to Blas is required to build/update the blas, so lifetimes get in the way sometimes.
Another problem is that acceleration structures don't keep there resources alive. (You get an error if you drop a vertex buffer used in a ac, when it is used).
I think it would be better (and more performant to validate), if it worked like Texture and TextureView, where the view keeps the texture alive if the Texture is dropped.

cwfitzgerald committed:
Add test -> check Dependency (#5033)

cwfitzgerald committed:
Dependabot Improvements (#5035)

dependabot[bot] committed:
Bump the patch-updates group with 11 updates (#5039)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

cwfitzgerald committed:
Pin rust-cache-version (#5040)

nical committed:
Move the trace recording of destroy events out of triage_resources (#5036)

nical committed:
Remove the free_resources tracker (#5037)

nical committed:
Remove the Destroyed state from Storage (#4970)

* Remove the Destroyed state from Storage

It used to be how we handled destroying buffers and textures but we moved to different approach.

* Explicit check for destroyed textures/buffers in a few entry points

This used to be checked automatically when getting the resource from the registry, but has to be done manually now that we track we track the destroyed state in the resources.

Wumpf committed:
Unify surface creation by introducing new `SurfaceTarget` enum (#4984)

nical committed:
Prevent a few integer overflows (#5042)

* Prevent a few integer overflows

* Add a changelog entry

* use u64

nical committed:
Triage suspected resources after triaging submissions (#5050)

cwfitzgerald committed:
Add proper timeouts to CI/nextest (#5051)

cwfitzgerald committed:
Convert Tests to Use Async Poll (#5053)

* Convert Tests to Use Async Poll

* Update examples/src/repeated_compute/mod.rs

Co-authored-by: Andreas Reich <r_andreas2@web.de>

---------

Co-authored-by: Andreas Reich <r_andreas2@web.de>

cwfitzgerald committed:
Fix DX12 Wait with Timeout (#5052)

daxpedda committed:
Add `cfg_aliases` to `wgpu-core` and `wgpu-hal` (#5055)

Commit:
Make DeviceLostClosure.from_c consume the closure before dropping it. (#5032)

This clarifies that the Rust and C-style callbacks/closures need to be
consumed (not called) before they are dropped. It also makes the from_c
function consume the param closure so that it can be dropped without
panicking.

It also relaxes the restriction that the callback/closure can only be
called once.

Wumpf committed:
Allow WebGPU &  WebGL in same wasm and detect WebGPU availability (#5044)

* Rename backends: web -> webgpu, direct -> wgpu_core

* rename context objects for web & core

* allow webgpu & webgl features side-by-side

* make sure webgl ci doesn't use webgpu

* update any_backend_feature_enabled

* add panicing generate_report method for compatibility

* RequestDeviceErrorKind::Web rename, fixup various cfg attributes

* automatic webgpu support detection

* changelog entry

* fix emscripten

* fix weird cfg, fix comment typo

* remove try_generate_report again

* Make get_mapped_range_as_array_buffer WebGPU only again

Vecvec committed:
Merge remote-tracking branch 'gfx-rs/trunk' into ray-tracing

# Conflicts:
#	wgpu-core/src/device/life.rs
#	wgpu/src/backend/wgpu_core.rs

Vecvec committed:
fix merge

Vecvec committed:
fix fmt

daniel-keitel commented on examples/src/ray_scene/mod.rs:
```diff
@@ -298,9 +298,8 @@ fn upload_scene_components(
-    load_model(&mut scene, "/examples/skybox/models/teslacyberv3.0.obj");
 
-    load_model(&mut scene, "/examples/ray-scene/cube.obj");
+    load_model(&mut scene, "/ray_scene/cube.obj");
```
```suggestion
    load_model(&mut scene, "/skybox/models/teslacyberv3.0.obj");
```

daniel-keitel commented on examples/src/ray_scene/mod.rs:
```diff
@@ -482,10 +475,14 @@ impl wgpu_example::framework::Example for Example {
                         .tlas_package
-                        .get_mut_single((x + y * side_count) as usize)
+                        .get_mut_single((x + y) * side_count)
                         .unwrap();
```
With this change the example should behave like before
```suggestion
                        .get_mut_single(x + y * side_count)
```
